code/artificial_intelligence/src/a_star/astar.cpp

```diff
@@ -0,0 +1,242 @@
+#include <list>
+#include <algorithm>
+#include <iostream>
+ 
+class point 
+{
+public:
+    point (int a = 0, int b = 0) 
+    { 
+        x = a; y = b; 
+    }
+
+    bool operator == (const point& o) 
+    { 
+        return o.x == x && o.y == y; 
+    }
+
+    point operator + (const point& o) 
+    { 
+        return point (o.x + x, o.y + y); 
+    }
+
+    int x, y;
+};
+ 
+class map 
+{
+public:
+    map() 
+    {
+        char t[8][8] = 
+        {
+            {0, 0, 0, 0, 0, 0, 0, 0}, {0, 0, 0, 0, 0, 0, 0, 0},
+            {0, 0, 0, 0, 1, 1, 1, 0}, {0, 0, 1, 0, 0, 0, 1, 0},
+            {0, 0, 1, 0, 0, 0, 1, 0}, {0, 0, 1, 1, 1, 1, 1, 0},
+            {0, 0, 0, 0, 0, 0, 0, 0}, {0, 0, 0, 0, 0, 0, 0, 0}
+        };
+
+        w = h = 8;
+        for (int r = 0; r < h; ++r)
+            for(int s = 0; s < w; ++s)
+                m[s][r] = t[r][s];
+    }
+
+    int operator() (int x, int y) 
+    { 
+        return m[x][y]; 
+    }
+
+    char m[8][8];
+    int w, h;
+};
+ 
+class node 
+{
+public:
+    bool operator == (const node& o) 
+    { 
+        return pos == o.pos; 
+    }
+
+    bool operator == (const point& o) 
+    {
+        return pos == o; 
+    }
+
+    bool operator < (const node& o) 
+    { 
+        return dist + cost < o.dist + o.cost; 
+    }
+
+    point pos, parent;
+    int dist, cost;
+};
+ 
+class aStar 
+{
+public:
+    aStar() 
+    {
+        neighbours[0] = point(-1, -1);
+        neighbours[1] = point(1, -1);
+        neighbours[2] = point(-1, 1);
+        neighbours[3] = point(1, 1);
+        neighbours[4] = point(0, -1);
+        neighbours[5] = point(-1, 0);
+        neighbours[6] = point(0, 1);
+        neighbours[7] = point(1, 0);
+    }
+ 
+    int calcDist(point& p)
+    {
+        // Adding a better heuristic
+        int x = end.x - p.x, y = end.y - p.y;
+        return(x * x + y * y);
+    }
+ 
+    bool isValid(point& p) 
+    {
+        return (p.x >-1 && p.y > -1 && p.x < m.w && p.y < m.h);
+    }
+ 
+    bool existPoint(point& p, int cost) 
+    {
+        std::list<node>::iterator i;
+        i = std::find(closed.begin(), closed.end(), p);
```

Comment:
Use `auto` and declare and define the variable on the same line. Iterators are typically called `it` as well.